pkg/storage/filepath/fs.go

```diff
@@ -186,23 +202,41 @@ func (fs *MemoryFS) Write(encoder runtime.Encoder, p string, obj runtime.Object,
 	fs.mu.Lock()
 	defer fs.mu.Unlock()
 
-	versionKey := filepath.Clean(p)
-	if version, ok := fs.versions[versionKey]; ok && version != storageVersion {
+	if rawObj, err := fs.readBuffer(p); err != nil {
+		if os.IsNotExist(err) {
+			// storageVersion == 0 -> this is a create, so it's expected to not exist (continue)
+			// storageVersion != 0 -> object has been deleted, propagate err to avoid a zombie update
+			if storageVersion != 0 {
+				return err
+			}
+		} else {
+			return err
+		}
+	} else if rawObj.version != storageVersion {
+		// this write is outdated
 		return VersionError
+	} else if bytes.Equal(rawObj.data, buf.Bytes()) {
+		// object serialized identically, skip write & version increment
+		return nil
 	}
 
 	dir, err := fs.ensureDir(filepath.Dir(p))
 	if err != nil {
 		return err
 	}
 
-	newVersion, err := resourceVersion(obj)
-	if err != nil {
+	// increment the resource version - it's applied to the object pointer for
+	// the caller in addition to being used to ensure the write is valid
+	newVersion := storageVersion + 1
```

Comment:
Ahhh I misunderstood and thought etcd was doing per-key revisions!

Switched back to global in e2ae723dae341170bd3b1da300c9ca8aa02c9a8f (though it now lives in `MemoryFS`)

Note that `RealFS` objects still have a per-object incrementing revision [serialized with the obj] so that they'll behave sanely across app restarts